File: 0076-minimum-window-substring/0076-minimum-window-substring.py
"""
Time = O(n)
Space = O(len(t))
"""
from collections import Counter
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(t) > len(s):
            return ""
        # Count each element in string t
        counter = Counter(t)
        need = len(t)
        shortest = ""
        shortest_len = len(s) + 1
        # Two pointers to create a interval
        l = 0
        for r in range(len(s)):
            # Check if the current element is in the dictionary
            if s[r] in counter:
                counter[s[r]] -= 1
                if counter[s[r]] >= 0:
                    need -= 1
            # Narrow down the interval
            while l <= r and need == 0:
                if r - l + 1 < shortest_len:
                    shortest = s[l: r + 1]
                    shortest_len = len(shortest)
                if s[l] in counter:
                    counter[s[l]] += 1
                    if counter[s[l]] > 0:
                        need += 1
                l += 1
        
        return shortest
                    
            

File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        size_t m = s.size(), n = t.size();
        if (m < n) {
            return "";
        }

        unordered_map<char, int> counter;
        int resStart = 0, resLen = INT_MAX;

        for (char c: t) {
            counter[c]++;
        }

        int left = 0, need = n;
        for (int right = 0; right < m; right++) {
            if (counter.find(s[right]) != counter.end()) {
                counter[s[right]]--;
                if (counter[s[right]] >= 0) {
                    need--;
                }
            }
            while (left <= right && need == 0) {
                if (right - left + 1 < resLen) {
                    resLen = right - left + 1;
                    resStart = left;
                }
                if (counter.find(s[left]) != counter.end()) {
                    counter[s[left]]++;
                    if (counter[s[left]] > 0) {
                        need++;
                    }
                }
                left++;
            }
        }
        if (resLen != INT_MAX) {
            return s.substr(resStart, resLen);
        } else {
            return "";
        }
    }
};